### PR TITLE
WT-5419 Add a TOMBSTONE to update on-disk KV pairs

### PR DESCRIPTION
**Pull request by kommiharibabu**
TOMBSTONE with WT_TS_NONE and WT_TXN_NONE to remove an on-disk key
and to remove an on-disk key remove operation, the TOMBSTONE update
with WT_TS_MAX and WT_TXN_MAX is added.

### CODE REVIEW

**Comment by kommiharibabu** [comment]
evergreen retry

**Comment by kommiharibabu** [comment]
evergreen retry

File: dist/s_clang-scan.diff [hunk 1]
```diff
@@ -14,14 +14,20 @@ src/conn/conn_capacity.c:291:5: warning: Value stored to 'capacity' is never rea
...
     ^          ~~~~~~~~~~~~~~~~~~
 1 warning generated.
-src/reconcile/rec_col.c:1050:23: warning: Null pointer passed as an argument to a 'nonnull' parameter
-                      memcmp(last.value->data, data, size) == 0)) {
-                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-1 warnings generated.
+src/reconcile/rec_col.c:1063:25: warning: Null pointer passed as an argument to a 'nonnull' parameter
+                        memcmp(last.value->data, data, size) == 0))) {
+                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+1 warning generated.
```
Comment (sueloverso) [inline dist/s_clang-scan.diff]: This is strange because your change did not modify `rec_col.c`. Why is this change in here?

File: dist/s_clang-scan.diff [hunk 2]
```diff
@@ -14,14 +14,20 @@ src/conn/conn_capacity.c:291:5: warning: Value stored to 'capacity' is never rea
...
 ./src/include/mutex.i:184:16: warning: Null pointer passed as an argument to a 'nonnull' parameter
     if ((ret = pthread_mutex_unlock(&t->lock)) != 0)
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1 warning generated.
+In file included from src/txn/txn_rollback_to_stable.c:9:
+In file included from ./src/include/wt_internal.h:421:
+./src/include/mutex.i:158:13: warning: Null pointer passed as an argument to a 'nonnull' parameter
+    return (pthread_mutex_trylock(&t->lock));
+            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+1 warning generated.
```
Comment (sueloverso) [inline dist/s_clang-scan.diff]: Again, you didn't change `wt_internal.h` nor did you change any locking near line 9 in `txn_rollback_to_stable.c` so I'm not sure why this changed. Admittedly you did change `txn_rollback_to_stable.c` and these errors are mysterious, but I'd still like to understand this.

File: src/txn/txn_rollback_to_stable.c [hunk 3]
```diff
@@ -126,6 +130,62 @@ __txn_abort_newer_insert(
         __txn_abort_newer_update(session, ins->upd, rollback_timestamp);
 }
 
+/*
+ * __txn_abort_fix_row_on_disk_kv --
+ *     Fix the on-disk row K/V version according to the given timestamp.
+ */
+static int
+__txn_abort_fix_row_on_disk_kv(
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW *rip, wt_timestamp_t rollback_timestamp)
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c]: I would avoid the use of `fix` in the function name here because it usually implies `fixed-length column store` everywhere else in WiredTiger. I suggest just removing that and using `txn_abort_row_ondisk_kv`.

**Comment by sueloverso** [comment]
I've made a few comments, but it is also not yet passing PR testing.

File: src/txn/txn_rollback_to_stable.c (hunk 3 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Changed the function name as per the suggestion.

File: src/txn/txn_rollback_to_stable.c [hunk 4]
```diff
@@ -110,6 +110,10 @@ __txn_abort_newer_update(
             upd->durable_ts = upd->start_ts = WT_TS_NONE;
         }
     }
+
+    /* Reset the history store flag for the stable update. */
+    if (first_upd)
+        first_upd->flags |= WT_UPDATE_HS;
```
Comment (keithbostic) [inline src/txn/txn_rollback_to_stable.c]: The function handles `first_upd == NULL`, but one of its callers checks before making the call. It seems that's inconsistent?

File: dist/s_clang-scan.diff (hunk 2 above)
Comment (kommiharibabu) [reply dist/s_clang-scan.diff]: Yes, I didn't change anything in `wt_internal.h` but a similar error like above is already masked, so I added this also.

File: dist/s_clang-scan.diff (hunk 1 above)
Comment (kommiharibabu) [reply dist/s_clang-scan.diff]: While updating the expected clang diff file, I generated the new diff and the line numbers are updated according to the code, so I update the line numbers.

File: src/txn/txn_rollback_to_stable.c [hunk 5]
```diff
@@ -126,6 +130,62 @@ __txn_abort_newer_insert(
...
+    /* Set the WT_UPDATE array reference. */
+    upd_entry = &mod->mod_row_update[WT_ROW_SLOT(page, rip)];
+    upd->next = NULL;
+    WT_ERR(__wt_update_serial(session, page, upd_entry, &upd, size, false));
+    WT_STAT_CONN_INCR(session, txn_rollback_upd_aborted);
+    return (0);
+
+err:
+    __wt_free(session, upd);
+    return (ret);
```
Comment (keithbostic) [inline src/txn/txn_rollback_to_stable.c]: Why isn't a column-store version needed?

File: src/txn/txn_rollback_to_stable.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: `upd` is used to check for the NULL condition in the above `for` loop. Once the loop is finished there should update that satisfy the given timestamp or there are no more updates. So NULL check is required before resetting the flag.

File: src/txn/txn_rollback_to_stable.c (hunk 5 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Column store is handled as part of another ticket, Once the complete row store is stabilized, column store is handled.

File: src/txn/txn_rollback_to_stable.c (hunk 4 above)
Comment (keithbostic) [reply src/txn/txn_rollback_to_stable.c]: I'm talking about the function itself -- some (but not all) of the callers check if they're passing a NULL as the `first_upd` argument to the function.

Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

File: src/txn/txn_rollback_to_stable.c [hunk 6]
```diff
@@ -123,7 +127,64 @@ __txn_abort_newer_insert(
...
+    WT_CELL_UNPACK *vpack, _vpack;
+    WT_DECL_RET;
+    WT_PAGE_MODIFY *mod;
+    WT_UPDATE *upd, **upd_entry;
+    size_t size;
+
+    vpack = &_vpack;
+    __wt_row_leaf_value_cell(session, page, rip, NULL, vpack);
+    if (vpack->start_ts > rollback_timestamp) {
+        /* FIXME: Get a visible update from the history store if any. */
```
Comment (agorrod) [inline src/txn/txn_rollback_to_stable.c]: Do you want to do this before or after merging this branch?

Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: I want to fix it before merging this branch.

File: src/txn/txn_rollback_to_stable.c [hunk 7]
```diff
@@ -123,7 +127,64 @@ __txn_abort_newer_insert(
...
+        upd->start_ts = WT_TS_NONE;
+    } else if (vpack->stop_ts != WT_TS_MAX && vpack->stop_ts > rollback_timestamp) {
+        /*
+         * Clear the remove operation from the key by inserting a TOMBSTONE with infinite
+         * visibility.
+         */
+        WT_RET(__wt_update_alloc(session, NULL, &upd, &size, WT_UPDATE_TOMBSTONE));
+        upd->txnid = WT_TXN_MAX;
+        upd->durable_ts = WT_TS_MAX;
+        upd->start_ts = WT_TS_MAX;
```
Comment (michaelcahill) [inline src/txn/txn_rollback_to_stable.c]: This part looks odd to me.

The on-disk value has a stop timestamp that is unstable, so a remove needs to be rolled back.  I would expect to see a standard update inserted containing the value and start time pair from on disk.  Maybe reconciliation will see this future tombstone and update the on-disk stop times, but it seems unnatural.

Put another way, if a value was removed with `commit_ts=100`, we wouldn't expect it to be removed again with `commit_ts=MAX`, so relying on reconciliation to do "the right thing" with a tombstone on top of a removed value seems fragile.

Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed by using the on-disk as a standard update.

File: src/reconcile/rec_visibility.c [hunk 8]
```diff
@@ -388,6 +388,11 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
               vpack->start_ts <= upd_select->stop_ts && vpack->start_txn <= upd_select->stop_txn);
             upd_select->durable_ts = upd_select->start_ts = vpack->start_ts;
             upd_select->start_txn = vpack->start_txn;
+
+            /* Use stop timestamp as durable timestamp if exist. */
+            if (upd_select->stop_ts != WT_TS_MAX)
+                upd_select->durable_ts = upd_select->stop_ts;
+
```
Comment (michaelcahill) [inline src/reconcile/rec_visibility.c]: It isn't obvious to me that the update should be made durable at the same timestamp it was removed.  Can you explain this part, preferably in the comment above the code?

File: src/txn/txn_rollback_to_stable.c [hunk 9]
```diff
@@ -111,6 +111,10 @@ __txn_abort_newer_update(
             upd->durable_ts = upd->start_ts = WT_TS_NONE;
         }
     }
+
+    /* Reset the history store flag for the stable update. */
+    if (first_upd != NULL)
+        F_CLR(first_upd, WT_UPDATE_HS);
```
Comment (michaelcahill) [inline src/txn/txn_rollback_to_stable.c]: Is this to make sure the stable update is written to history?  Can there be more updates with the flag set, and if so, is that okay?  Ideally this would be explained in the comment: at the moment it is describing what the code does, but not why.

File: src/txn/txn_rollback_to_stable.c (hunk 6 above)
Comment (michaelcahill) [reply src/txn/txn_rollback_to_stable.c]: Does that mean this branch isn't ready for review?

File: src/txn/txn_rollback_to_stable.c [hunk 10]
```diff
@@ -199,7 +271,13 @@ __txn_abort_newer_row_leaf(
 
         if ((insert = WT_ROW_INSERT(page, rip)) != NULL)
             __txn_abort_newer_insert(session, insert, rollback_timestamp);
+
+        /* Check for updates in the update list that satisfy the given timestamp. */
+        if (WT_ROW_UPDATE(page, rip) == NULL && WT_ROW_INSERT(page, rip) == NULL)
+            WT_RET(__txn_abort_row_ondisk_kv(session, page, rip, rollback_timestamp));
```
Comment (michaelcahill) [inline src/txn/txn_rollback_to_stable.c]: Could there be updates needing rollback, and the on-disk value *also* needs rollback?  I'm not clear why the `if` is needed around this call.

File: src/reconcile/rec_visibility.c (hunk 8 above)
Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c]: I refactored this change to use the durable timestamp of the TOMBSTONE record. With this change, it uses the proper durable timestamp.

File: src/txn/txn_rollback_to_stable.c (hunk 9 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: We are clearing the flag to indicate that this update is not moved into the history store. I added a comment to explain. Is that comment provides better understanding?

File: src/txn/txn_rollback_to_stable.c (hunk 10 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Yes, the check is not required. But this can be used for optimizing for the cases where there is a stable update is present, so there is no need to abort the on-disk value even if it is written a newer version, as later in the reconciliation, that will be updated. I am doing this change in the next pull request.

**Comment by kommiharibabu** [comment]
evergreen retry

File: src/reconcile/rec_visibility.c [hunk 11]
```diff
@@ -368,9 +371,9 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
                 upd_select->start_txn = upd->txnid;
             }
 
-            /* Use stop timestamp as durable timestamp if exist. */
-            if (upd_select->stop_ts != WT_TS_MAX)
-                upd_select->durable_ts = upd_select->stop_ts;
+            /* Use tombstone durable timestamp as overall durable timestamp if exist. */
```
Comment (sueloverso) [inline src/reconcile/rec_visibility.c]: I suggest rewording as: "Use the tombstone durable timestamp as the overall durable timestamp if it exists."

File: src/reconcile/rec_visibility.c [hunk 12]
```diff
@@ -388,6 +391,11 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
               vpack->start_ts <= upd_select->stop_ts && vpack->start_txn <= upd_select->stop_txn);
             upd_select->durable_ts = upd_select->start_ts = vpack->start_ts;
             upd_select->start_txn = vpack->start_txn;
+
+            /* Use tombstone durable timestamp as overall durable timestamp if exist. */
```
Comment (sueloverso) [inline src/reconcile/rec_visibility.c]: Same suggestion here.

File: src/txn/txn_rollback_to_stable.c [hunk 13]
```diff
@@ -111,6 +111,15 @@ __txn_abort_newer_update(
             upd->durable_ts = upd->start_ts = WT_TS_NONE;
         }
     }
+
+    /*
+     * Reset the history store flag for the stable update to indicate that this update is not
+     * written into history store, later all the aborted updates are removed from history store.
+     * Next time when this update is moved into history store, it will have different stop time
+     * pair.
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c]: Minor rewording suggestion:
Clear the history store flag for the stable update to indicate that this update should not be
written into the history store later, when all the aborted updates are removed from the history store.
The next time when this update is moved into the history store, it will have a different stop time
pair.

File: src/txn/txn_rollback_to_stable.c [hunk 14]
```diff
@@ -124,7 +133,73 @@ __txn_abort_newer_insert(
...
+    WT_DECL_RET;
+    WT_ITEM buf;
+    WT_PAGE_MODIFY *mod;
+    WT_UPDATE *upd, **upd_entry;
+    size_t size;
+
+    vpack = &_vpack;
+    __wt_row_leaf_value_cell(session, page, rip, NULL, vpack);
+    if (vpack->start_ts > rollback_timestamp) {
+        /* No visible updates in the history store, may be a newly inserted key. Remove it. */
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c]: The comment is hard to understand. The part about "may be a newly inserted key" leads me to wonder what else this key could be. I suggest simplifying with:
There are no visible updates in the history store for this key. Remove it.

File: src/txn/txn_rollback_to_stable.c [hunk 15]
```diff
@@ -124,7 +133,73 @@ __txn_abort_newer_insert(
...
+        upd->start_ts = WT_TS_NONE;
+    } else if (vpack->stop_ts != WT_TS_MAX && vpack->stop_ts > rollback_timestamp) {
+        /*
+         * Clear the remove operation from the key by inserting the original on-disk value as a
+         * standard update.
+         */
+        WT_CLEAR(buf);
+
+        /*
+         * If a value is simple and is globally visible at the time of reading a page into cache, we
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c]: I do not understand the definition of "simple" in this comment. Does it mean not an overflow record? Something else?

File: src/txn/txn_rollback_to_stable.c [hunk 16]
```diff
@@ -199,7 +274,12 @@ __txn_abort_newer_row_leaf(
 
         if ((insert = WT_ROW_INSERT(page, rip)) != NULL)
             __txn_abort_newer_insert(session, insert, rollback_timestamp);
+
+        /* Abort any on-disk value */
```
Comment (sueloverso) [inline src/txn/txn_rollback_to_stable.c]: End comment sentence with a period.

File: test/suite/test_rollback_to_stable01.py [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+        self.check(valuea, uri, 0, 20)
+
+        # Pin oldest and stable to timestamp 100.
+        self.conn.set_timestamp('stable_timestamp=' + timestamp_str(10))
+
+        # Checkpoint to ensure that all the updates are flushed to disk.
+        self.session.checkpoint()
+
+        self.conn.rollback_to_stable()
+
```
Comment (sueloverso) [inline test/suite/test_rollback_to_stable01.py]: This test probably does not need a blank line after every single line. Please review and remove some.

File: src/reconcile/rec_visibility.c (hunk 11 above)
Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c]: Fixed as suggested.

File: src/reconcile/rec_visibility.c (hunk 12 above)
Comment (kommiharibabu) [reply src/reconcile/rec_visibility.c]: Fixed as suggested.

File: src/txn/txn_rollback_to_stable.c (hunk 13 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed as suggested.

File: src/txn/txn_rollback_to_stable.c (hunk 14 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

File: src/txn/txn_rollback_to_stable.c (hunk 16 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Fixed.

File: test/suite/test_rollback_to_stable01.py (hunk 17 above)
Comment (kommiharibabu) [reply test/suite/test_rollback_to_stable01.py]: Fixed by removing newlines where they are not necessary.

File: src/txn/txn_rollback_to_stable.c (hunk 15 above)
Comment (kommiharibabu) [reply src/txn/txn_rollback_to_stable.c]: Simple means the values that are not compressed, I updated the comment to explain it. Those values when they are globally visible, they will be placed into the WT_ROW itself to directly reference it.

**Comment by kommiharibabu** [comment]
The failure is as expected and it will be fixed once the transaction support is removed from hs insert/remove.
